ARCHITECTURE.md

```diff
@@ -0,0 +1,207 @@
+      * OTOH, you don't get to "use Haskell because everything else should be considered lame and fundamentally broken" 😉.
+
+</details>
+
+### Tenet 5: “The Rails Way” (vs. overfitting to today’s needs)
+
+Prefer the conventional Rails way to accomplish a task: it might be better to be 100% conventional than get a 10% gain by deviating from convention. This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.
+
+<details>
+  <summary>Why?</summary>
+
+  * Observing the codebase over a decade of dev: sometimes you can get a gain this year by customizing Rails to work a different way that’s a little better for us, but over time non-conventional code tends to become a maintenance burden.  
+  * Often the “10% better” non-conventional code is exactly what prevents us from upgrading to the “50% better” official Rails solution to the problem that comes out 2 years later.  
+  * By and large we have pretty conventionally Rails-shaped problems, which means Rails tends to solve our problems better than we would if we wait a little longer.  
+  * Rails has been a great long-term fit to our problem space and we benefit from continuing to get upgrades as Rails has improved. Doing it “their way” makes upgrades happen semi-automatically.
+
+</details>
+
+### Tenet 6: Managed Services that implement an Open Source Protocol / API (vs proprietary managed services or self-hosted open source services)
+
+When making technical selections, prefer managed services that implement an Open Protocol or an Open Source API. For example, we don’t host our own MySQL instance, but we ALSO don’t use ProprietarySQLDB(TM). Instead, we use a 3rd party managed service that implements the MySQL protocol.
+
+<details>
+  <summary>Why?</summary>
+
+  * Gives us the best of both worlds:
+    * Self-hosting many open source servers, especially stateful services like DBs, requires ongoing infra admin work and monitoring attention. We have a small team.  
+    * Developing against an API with an Open Source implementation gives us flexibility:  
+      * …to implement a “local dev” for the service using the open source version  
+      * …to change vendors if need be (cost, availability, in-kind donations, etc)  
+
+</details>
+
+### Tenet 7: Use REST (vs. GraphQL) for APIs
+
+As we add APIs for our platform, we will publish them using REST. Client-side apps should communicate with the server using REST endpoints.
+
+<details>
+  <summary>Why?</summary>
+
+  * KISS: REST has been a simple and an effective match to our needs.  
+  * Our labs and other fat client-side JS apps don’t tend to have the “complex aggregation of dynamic data from many sources” property that benefits from more complex systems like GraphQL.  
+  * Memoization is easily applied to REST-based calls (good for hosting cost and at-scale events like hour of code).
```

Comment:
it feels like cacheability should be on this list -- REST GET requests are cacheable by cloudfront / firewalls / browsers while GraphQL requests generally are not. one possible phrasing:

```
REST APIs inherently support HTTP caching mechanisms (e.g., ETag, Last-Modified, Cache-Control), enabling better performance and reducing load on the server.
```

separately, I'm not sure what you mean by memoization -- if you do mean caching, I think we should call it caching or cacheability. if not, can you say more about what memoization means in this case?

Reply:
I inherited that line from the previous doc, and also agree that I'd rather write it more simply as supporting caching.

Reply:
changed